hooli_data_eng/assets/dbt_assets.py

```diff
@@ -246,3 +253,56 @@ def dbt_slim_ci(dbt2: DbtCliResource):
 @job
 def dbt_slim_ci_job():
     dbt_slim_ci()
+
+
+# Load and parse the dbt manifest
+def load_dbt_manifest(manifest_path):
+    with open(manifest_path, 'r') as f:
+        manifest = json.load(f)
+    return manifest
+
+# Create AssetCheckSpec objects from the dbt manifest
+def create_source_freshness_check_specs(manifest_path):
+    manifest = load_dbt_manifest(manifest_path)
+    check_specs = []
+    for _, source_data in manifest['sources'].items():
+        if "freshness" in source_data and source_data["freshness"].get("error_after").get("count") is not None and source_data["name"] != "locations":
+            asset_key = AssetKey([source_data["schema"].upper(), source_data["name"] ])
+            check_specs.append(
+                AssetCheckSpec(
+                    name="source_freshness_check",
+                    asset=asset_key,
+                    blocking=True,
+                    automation_condition=allow_outdated_and_missing_parents_condition),
+            )
+    return check_specs
+
+@multi_asset_check(
+    specs=create_source_freshness_check_specs(dbt_project.manifest_path),
+    can_subset=True,
+    required_resource_keys={"dbt"},
+)
+def source_freshness_check(context: AssetCheckExecutionContext):
+    dbt_resource = context.resources.dbt
+
+    selection_list = [f"source:{asset_check_key.asset_key.path[0].lower()}.{asset_check_key.asset_key.path[1]}" for asset_check_key in context.selected_asset_check_keys]
+    selection_string = " ".join(selection_list)
+
+    context.log.info(f"Running check source_freshness_check on assets {selection_string}")
+    dbt_cli_invocation = dbt_resource.cli(["source", "freshness", "--select", selection_string], raise_on_error=False)
+    # TODO add better metadata
+    # Process the raw event stream
+    for dbt_event in dbt_cli_invocation.stream_raw_events():
+        context.log.info(f"Received event: {dbt_event.raw_event}")
+        if dbt_event.raw_event.get('info', {}).get('name') == 'LogFreshnessResult':
+            log_freshness_result = dbt_event.raw_event['info']
+            context.log.info(f"Filtered LogFreshnessResult: {log_freshness_result}")
+            passed = True if log_freshness_result['level'] == 'info' else False
+            severity = AssetCheckSeverity.ERROR if log_freshness_result['level'] == 'error' else AssetCheckSeverity.WARN
+            yield AssetCheckResult(
+                asset_key=AssetKey([dbt_event.raw_event['data']['node_info']['node_relation']['schema'].upper(), dbt_event.raw_event['data']['node_info']['node_name']]),
```

Comment:
one nit: there is a lot asset key to dbt name transformations going on in this PR, and a nice refactor might be to pull those out into separate functions

Reply:
Will do. Another thing I didn't love about this was putting asset checks inside `dbt_asset.py`. Since I'm removing the automation condition it would be straightforward to move this into a separate file (or at least the supporting functions).

Reply:
on second look, none of these are repeated -- do you think it makes sense to have one-off functions for each of these use cases?

Reply:
Ultimately up to you, I think functions would help future us remember why we're doing all this .upper / .lower stuff lol, but I can also buy the case that you could figure it out reasonably well from the context